keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -99,105 +101,152 @@ A developer has a Windows web app that they would like to deploy in a container
-  - Create a role to view, modify or use the GMSA configmap:
+  - Create a cluster wide role to get and "use" the GMSACredSpec:
 
 ```
-kind: Role
 apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
 metadata:
-  namespace: default
-  name: webserver-gmsa-user
+  name: webapp1-gmsa-user
 rules:
-- apiGroups: [""]
-  resources: ["configmaps"]
-  resourceNames: ["webserver-credspec"]
-  verbs: ["get", "patch", "delete", "update", "use"]
+- apiGroups: ["auth.k8s.io"]
+  resources: ["gmsacredspecs"]
+  resourceNames: ["webapp1-credspec"]
+  verbs: ["get", "use"]
 ```
 
-  - Create a rolebinding and assign desired users (for example, jane below) or service accounts to the above role:
+  - Create a rolebinding and assign desired service accounts to the above role:
 
 ```
-kind: RoleBinding
 apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
 metadata:
-  name: use-webserver-gmsa
+  name: use-webapp1-gmsa
   namespace: default
 subjects:
-- kind: User
-  name: jane
+- kind: ServiceAccount
+  name: webapp-sa
   apiGroup: rbac.authorization.k8s.io
 roleRef:
   kind: Role
-  name: webserver-gmsa-user
+  name: webapp1-gmsa-user
   apiGroup: rbac.authorization.k8s.io
 ```
   
-4. Application admins deploy app pods with a reference to the name of an authorized GMSA configmap (e.g. `webserver-credspec`). In the alpha and beta stage of this feature, the reference to the GMSA configmap will be set through an annotation on the pod: `pod.[alpha|beta].kubernetes.io/windows-gmsa-config-map`. In the beta and GA stage, the reference to GMSA configmap may be set through `spec.securityContext.windows.credentialSpecConfig` in the pod spec.
-5. During pod creation, a new Admission Controller, GMSAAuthorizer, will ensure the `user` (and `spec.serviceAccountName` if specified) is authorized for the `use` verb on the GMSA configmap. If the authorization check fails due to absence of RBAC roles, the pod creation fails.
-6. The Windows CRI implementation accesses the configmap with the GMSA credspec and copies the contents of entry with key `gmsa-cred-spec` to the [OCI windows.CredentialSpec](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#credential-spec) field.
-7. The Windows OCI implementation validates the credspec and fails to start the container if its invalid or access to the GMSA from the node is denied.
-8. The container starts with the Windows Group Managed Service Account as expected, and can authenticate to the database.
-9. During any subsequent pod update, any changes to a pod's `securityContext` is blocked (as is the case today) by `ValidatePodUpdate`.
+5. Application admins will deploy app pods that require a GMSA identity along with a Service Account authorized to use the GMSAs. There will be two options regarding how to specify the GMSACredSpec for pods and containers:
```

Comment:
Why do we want both options? Would it be less confusing to users to just stick with one for now?

Reply:
I am calling out both options here for completeness. As called out towards the end of the bullet, users will typically just use the first option which is more user friendly. The second option (inline/expanded form) is an artifact of the design suggestion to keep the CRI changes simple and avoid having to look up objects in a much later stage of the pod creation process and potentially failing there along with other access control concerns. These are detailed in the section `Specifying only the name of GMSACredSpec objects in pod spec fields/annotations` under `Alternatives`